### PR TITLE
Dijkstra algorithm

### PR DESCRIPTION
**Pull request by brandesdavid**
# Pull Request Template

## Description

added a dijkstra algorithm

Fixes: #60 

## Type of change

- [X] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

get nearest stations and test the distances of some paths

## Checklist:

Before submitting your PR, please review the following checklist:

- [X] I have performed a self-review of my own code or materials.
- [X] I have commented my code, particularly in hard-to-understand areas.
- [X] I have made corresponding changes to the documentation.
- [X] My changes generate no new warnings.
- [X] I have added tests that prove my fix is effective or that my feature works.
- [X] New and existing unit tests pass locally with my changes.
- [X] Any dependent changes have been merged and published in downstream modules.

## Additional Information:

woah this was a beast. we should take a pause after that

### CODE REVIEW

File: api/getStationDistance.go [hunk 1]
```diff
@@ -1,183 +1,211 @@
...
 
-var httpClient = &http.Client{}
+var stationsList map[string]utils.StationListEntry
+var stationIDs []string
+var linesList map[string][]string
 
-// parsing the float from the query parameters, which are strings
-func parseFloat(str string) (float64, error) {
-	val, err := strconv.ParseFloat(str, 64)
+func SetupStationsListAndLinesList() (map[string]utils.StationListEntry, []string, map[string][]string) {
```
Comment (johan-t) [inline api/getStationDistance.go:25]: Can you find a name for this function that explains how it is setting it up or what exactly it does?

File: api/getStationDistance.go [hunk 2]
```diff
@@ -1,183 +1,211 @@
...
+			queue.PushBack(StationNode{adjacentStationId, newDistance, stationsList[adjacentStationId].Lines[0]})
+		}
 	}
+}
+
+// this is the dijkstra algorithm
+func FindShortestDistance(startStation string, userLat, userLon float64) int {
+	SetupStationsListAndLinesList()
+
+	//find the nearest station to the user (returns the station ID)
```
Comment (johan-t) [inline api/getStationDistance.go:175]: You can remove the comment by specifying the type and giving the function a better name

**Review by johan-t — CHANGES_REQUESTED** [review]
some style changes needed and a change to the api. Rest looks fantastic

File: utils/dijkstraUtilities.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+			math.Sin(dLon/2)*
+			math.Sin(dLon/2)
+
+	c := 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
+
+	distance := R * c // Distance in km
+	return distance
+}
+
+// with this function, we can get the next and previous station of our current station in a line
```
[the remark this answers is not in the document]
Comment (brandesdavid) [reply utils/dijkstraUtilities.go:35]: yes

File: README.md [hunk 4]
```diff
@@ -230,26 +230,33 @@ It will return the station.json with this body:
 
 ### Get the station distance of two locations
 
-- `/distance` - This endpoint is used to get the station distance in stops between two locations
+- `/distance` - This endpoint is used to get the station distance in stops between two location
+
+#### Process:
+  1. We get the coordinates from our inspector's station id 
+  2. Calculate the distance of the user and the station, if it is less than 1 km, we return 1 station distance
```
[the remark this answers is not in the document]
Comment (brandesdavid) [reply README.md:237]: https://github.com/FreiFahren/frontend/issues/94

**Pull request merged**